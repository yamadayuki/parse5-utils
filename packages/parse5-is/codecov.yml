coverage:
  status:
    project:
      is:
        flags: is
    patch:
      is:
        flags: is

fixes:
  - "::src/packages/parse5-is"
