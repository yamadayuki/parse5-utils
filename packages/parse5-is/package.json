{
  "name": "@yamadayuki/parse5-is",
  "version": "0.2.1",
  "description": "check a passed `parse5` AST node is what",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "prebuild": "npm run clean",
    "build": "tsc -p .",
    "build:watch": "tsc -p . -w",
    "clean": "rm -rf lib",
    "test": "jest -c ./jest.config.js",
    "test:update": "jest -c ./jest.config.js -u",
    "coverage": "jest -c ./jest.config.js --collectCoverage"
  },
  "keywords": [],
  "author": "Yuki Yamada <yamadayuuki0818@gmail.com>",
  "repository": "https://github.com/yamadayuki/parse5-utils/tree/master/packages/parse5-is",
  "publishConfig": {
    "access": "public"
  },
  "license": "MIT",
  "devDependencies": {
    "@types/jest": "^25.2.3",
    "@types/node": "^12.0.0",
    "@types/parse5": "^5.0.0",
    "jest": "^24.1.0",
    "parse5": "^5.1.0",
    "ts-jest": "^24.0.0",
    "typescript": "^3.2.2"
  }
}
