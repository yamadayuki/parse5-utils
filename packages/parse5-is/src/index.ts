import {
  DefaultTreeCommentNode,
  DefaultTreeDocument,
  DefaultTreeDocumentFragment,
  DefaultTreeDocumentType,
  DefaultTreeElement,
  DefaultTreeNode,
  DefaultTreeTextNode,
  DefaultTreeParentNode,
} from "parse5";

export function isElement(node: DefaultTreeNode): node is DefaultTreeElement {
  return node.nodeName !== undefined && node.nodeName !== null;
}

export function isDocument(node: DefaultTreeNode): node is DefaultTreeDocument {
  return node.nodeName === "#document";
}

export function isTextNode(node: DefaultTreeNode): node is DefaultTreeTextNode {
  return node.nodeName === "#text";
}

export function isCommentNode(
  node: DefaultTreeNode
): node is DefaultTreeCommentNode {
  return node.nodeName === "#comment";
}

export function isDocumentFragment(
  node: DefaultTreeNode
): node is DefaultTreeDocumentFragment {
  return node.nodeName === "#document-fragment";
}

export function isDocumentType(
  node: DefaultTreeNode
): node is DefaultTreeDocumentType {
  return node.nodeName === "#documentType";
}

export function hasSourceCodeLocation(
  node: DefaultTreeElement | DefaultTreeTextNode
): node is Required<typeof node> {
  return node.sourceCodeLocation !== undefined;
}

export function hasParentNode(
  node: DefaultTreeNode
): node is DefaultTreeNode & { parentNode: DefaultTreeParentNode } {
  return !(
    isDocument(node) ||
    isDocumentFragment(node) ||
    isDocumentType(node)
  );
}

export function hasChildNodes(
  node: DefaultTreeNode
): node is DefaultTreeNode & DefaultTreeParentNode {
  return isDocument(node) || isDocumentFragment(node) || isElement(node);
}
