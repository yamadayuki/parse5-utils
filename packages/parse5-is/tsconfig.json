{
  "extends": "../../tsconfig.base",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "include": ["./src/**/*"],
  "exclude": ["**/__tests__/*"]
}
