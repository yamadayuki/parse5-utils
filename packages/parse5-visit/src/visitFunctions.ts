import {
  hasChildNodes,
  isCommentNode,
  isDocument,
  isDocumentFragment,
  isDocumentType,
  isElement,
  isTextNode,
} from "@yamadayuki/parse5-is";
import { CommentNode, Document, DocumentFragment, DocumentType, Element, Node } from "parse5";
import { VisitorFunction } from "./types";

export function visitDocument(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<Document>;
    onLeave?: VisitorFunction<Document>;
  }
) {
  if (isDocument(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    if (node.childNodes && node.childNodes.length > 0) {
      node.childNodes.forEach(childNode => {
        visitDocument(childNode, { onEnter, onLeave });
      });
    }
  }

  if (isDocument(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}

export function visitDocumentFragment(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<DocumentFragment>;
    onLeave?: VisitorFunction<DocumentFragment>;
  }
) {
  if (isDocumentFragment(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    node.childNodes.forEach(childNode => {
      visitDocumentFragment(childNode, { onEnter, onLeave });
    });
  }

  if (isDocumentFragment(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}

export function visitDocumentType(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<DocumentType>;
    onLeave?: VisitorFunction<DocumentType>;
  }
) {
  if (isDocumentType(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    if (node.childNodes && node.childNodes.length > 0) {
      node.childNodes.forEach(childNode => {
        visitDocumentType(childNode, { onEnter, onLeave });
      });
    }
  }

  if (isDocumentType(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}

export function visitElement(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<Element>;
    onLeave?: VisitorFunction<Element>;
  }
) {
  if (isElement(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    node.childNodes.forEach(childNode => {
      visitElement(childNode, { onEnter, onLeave });
    });
  }

  if (isElement(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}

export function visitCommentNode(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<CommentNode>;
    onLeave?: VisitorFunction<CommentNode>;
  }
) {
  if (isCommentNode(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    node.childNodes.forEach(childNode => {
      visitCommentNode(childNode, { onEnter, onLeave });
    });
  }

  if (isCommentNode(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}

export function visitTextNode(
  node: Node,
  {
    onEnter,
    onLeave,
  }: {
    onEnter?: VisitorFunction<Node>;
    onLeave?: VisitorFunction<Node>;
  }
) {
  if (isTextNode(node) && typeof onEnter === "function") {
    onEnter(node);
  }

  if (hasChildNodes(node)) {
    node.childNodes.forEach(childNode => {
      visitTextNode(childNode, { onEnter, onLeave });
    });
  }

  if (isTextNode(node) && typeof onLeave === "function") {
    onLeave(node);
  }

  return node;
}
