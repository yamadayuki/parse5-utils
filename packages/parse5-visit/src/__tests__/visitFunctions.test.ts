import { hasParentNode } from "@yamadayuki/parse5-is";
import { DefaultTreeDocument, DefaultTreeDocumentFragment, DefaultTreeElement, parse, parseFragment } from "parse5";
import { visitDocument, visitDocumentFragment, visitElement } from "../visitFunctions";

describe("visitDocument", () => {
  const html = `<!DOCTYPE html>
  <html>
  <body>
    <h1>My First Heading</h1>
    <p>My first paragraph.</p>
  </body>
  </html>`;
  const parsed = parse(html);

  it("doesn't throw", () => {
    const visitor = jest.fn((node: DefaultTreeDocument) => node);

    expect(() => {
      visitDocument(parsed as DefaultTreeDocument, { onEnter: visitor });
    }).not.toThrow();
  });

  it("calls `onEnter` only once", () => {
    const visitor = jest.fn((node: DefaultTreeDocument) => node);

    visitDocument(parsed as DefaultTreeDocument, { onEnter: visitor });
    /**
     * the visitor is called only one time in this suite.
     * #document <- call!
     *   html
     *     head
     *     body
     *       h1
     *       p
     */
    expect(visitor).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const affectedNodeName: string[] = [];
    const onEnter = jest.fn((node: DefaultTreeDocument) => {
      affectedNodeName.push(`-> ${node.nodeName}`);
      return node;
    });
    const onLeave = jest.fn((node: DefaultTreeDocument) => {
      affectedNodeName.push(`<- ${node.nodeName}`);
      return node;
    });

    visitDocument(parsed as DefaultTreeDocument, { onEnter, onLeave });
    expect(affectedNodeName).toMatchSnapshot();
  });
});

describe("visitDocumentFragment", () => {
  const html = `<!DOCTYPE html>
  <html>
  <body>
    <h1>My First Heading</h1>
    <p>My first paragraph.</p>
  </body>
  </html>`;
  const parsed = parseFragment(html);

  it("doesn't throw", () => {
    const visitor = jest.fn((node: DefaultTreeDocumentFragment) => node);

    expect(() => {
      visitDocumentFragment(parsed as DefaultTreeDocumentFragment, { onEnter: visitor });
    }).not.toThrow();
  });

  it("calls `onEnter` only once", () => {
    const visitor = jest.fn((node: DefaultTreeDocumentFragment) => node);

    visitDocumentFragment(parsed as DefaultTreeDocumentFragment, { onEnter: visitor });
    /**
     * the visitor is called only one time in this suite.
     * #document-fragment <- call!
     *   h1
     *   p
     */
    expect(visitor).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const affectedNodeName: string[] = [];
    const onEnter = jest.fn((node: DefaultTreeDocumentFragment) => {
      affectedNodeName.push(`-> ${node.nodeName}`);
      return node;
    });
    const onLeave = jest.fn((node: DefaultTreeDocumentFragment) => {
      affectedNodeName.push(`<- ${node.nodeName}`);
      return node;
    });

    visitDocumentFragment(parsed as DefaultTreeDocumentFragment, { onEnter, onLeave });
    expect(affectedNodeName).toMatchSnapshot();
  });
});

describe("visitElement", () => {
  const html = `<!DOCTYPE html>
  <html>
  <body>
    <h1>My First Heading</h1>
    <p>My first paragraph.</p>
  </body>
  </html>`;
  const parsed = parse(html);

  it("doesn't throw", () => {
    const visitor = jest.fn((node: DefaultTreeElement) => node);

    expect(() => {
      visitElement(parsed as DefaultTreeElement, { onEnter: visitor });
    }).not.toThrow();
  });

  it("calls `onEnter` only once", () => {
    const visitor = jest.fn((node: DefaultTreeElement) => node);

    visitElement(parsed as DefaultTreeElement, { onEnter: visitor });
    /**
     * the visitor is called only one time in this suite.
     * #document
     *   html    <- call!
     *     head  <- call!
     *     body  <- call!
     *       h1  <- call!
     *       p   <- call!
     */
    expect(visitor).toHaveBeenCalledTimes(5);
  });

  it("matches snapshot", () => {
    const affectedNodeName: string[] = [];
    let depth = 0;
    const onEnter = jest.fn((node: DefaultTreeElement) => {
      if (hasParentNode(node)) {
        depth = depth + 1;
      }
      affectedNodeName.push(`${"".padStart(depth * 2, " ")}-> ${node.nodeName}`);
      return node;
    });
    const onLeave = jest.fn((node: DefaultTreeElement) => {
      affectedNodeName.push(`${"".padStart(depth * 2, " ")}<- ${node.nodeName}`);
      if (hasParentNode(node)) {
        depth = depth - 1;
      }
      return node;
    });

    visitElement(parsed as DefaultTreeElement, { onEnter, onLeave });
    expect(affectedNodeName).toMatchSnapshot();
  });
});
