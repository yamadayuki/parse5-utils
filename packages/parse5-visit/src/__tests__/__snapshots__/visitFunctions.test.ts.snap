// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitCommentNode matches snapshot 1`] = `
Array [
  "  -> #comment",
  "  <- #comment",
]
`;

exports[`visitDocument matches snapshot 1`] = `
Array [
  "-> #document",
  "<- #document",
]
`;

exports[`visitDocumentFragment matches snapshot 1`] = `
Array [
  "-> #document-fragment",
  "<- #document-fragment",
]
`;

exports[`visitDocumentType matches snapshot 1`] = `
Array [
  "-> #documentType",
  "<- #documentType",
]
`;

exports[`visitElement matches snapshot 1`] = `
Array [
  "  -> html",
  "    -> head",
  "    <- head",
  "    -> body",
  "      -> h1",
  "      <- h1",
  "      -> p",
  "      <- p",
  "    <- body",
  "  <- html",
]
`;

exports[`visitTextNode matches snapshot 1`] = `
Array [
  "  -> #text",
  "  <- #text",
  "  -> #text",
  "  <- #text",
]
`;
